packages/compass-schema-validation/src/modules/validation.ts

```diff
@@ -144,280 +170,271 @@ export const checkValidator = (
- */
-const setValidation = (
-  state: ValidationState,
-  action: ValidationFetchedAction | ValidationCanceledAction
-): ValidationState => {
-  const checkedValidator = checkValidator(action.validation.validator ?? '');
-  // TODO(COMPASS-4989): javascriptStringify??
-  const validator = javascriptStringify(
-    checkedValidator.validator,
-    null,
-    2
-  ) as string;
-
-  return {
-    ...state,
-    prevValidation: {
-      validator,
-      validationAction: action.validation.validationAction,
-      validationLevel: action.validation.validationLevel,
-    },
-    isChanged: action.validation.isChanged || false,
-    validator: validator,
-    validationAction: action.validation.validationAction,
-    validationLevel: action.validation.validationLevel,
-    syntaxError: null,
-    error: action.validation.error || null,
-  };
-};
+  if (
+    isAction<ValidationSaveFailedAction>(
+      action,
+      ValidationActions.ValidationSaveFailed
+    )
+  ) {
+    return {
+      ...state,
+      error: action.error,
+    };
+  }
 
-/**
- * Set Error.
- */
-const setError = (
-  state: ValidationState,
-  action: ValidationSaveFailedAction
-): ValidationState => {
-  return {
-    ...state,
-    error: action.error || null,
-  };
-};
+  if (
+    isAction<EmptyValidationFetchedAction>(
+      action,
+      ValidationActions.EmptyValidationFetched
+    )
+  ) {
+    return {
+      ...state,
+      isChanged: false,
+      error: null,
+      syntaxError: null,
+      ...DEFAULT_VALIDATION,
+      validator: action.validationTemplate,
+    };
+  }
 
-/**
- * Change validation action.
- */
-const changeValidationAction = (
-  state: ValidationState,
-  action: ValidationActionChangedAction
-): ValidationState => {
-  const newState = {
-    ...state,
-    validationAction: action.validationAction,
-  };
+  if (
+    isAction<ValidationFetchErroredAction>(
+      action,
+      ValidationActions.ValidationFetchErrored
+    )
+  ) {
+    return {
+      ...state,
+      prevValidation: {
+        ...DEFAULT_VALIDATION,
+      },
+      isChanged: true,
+      error: action.error,
+      syntaxError: null,
+      ...DEFAULT_VALIDATION,
+    };
+  }
 
-  return {
-    ...newState,
-    isChanged: !isEqual(
-      pick(newState, ['validator', 'validationAction', 'validationLevel']),
-      state.prevValidation
-    ),
-  };
-};
+  if (
+    isAction<ValidationFetchedAction>(
+      action,
+      ValidationActions.ValidationFetched
+    )
+  ) {
+    // TODO(COMPASS-4989): EJSON??
+    const checkedValidator = checkValidator(
+      EJSON.stringify(action.validator, undefined, 2)
+    );
+    // TODO(COMPASS-4989): javascriptStringify??
+    const validator = javascriptStringify(
+      checkedValidator.validator,
+      null,
+      2
+    ) as string;
+
+    return {
+      ...state,
+      prevValidation: {
+        validator,
+        validationAction: action.validationAction,
+        validationLevel: action.validationLevel,
+      },
```

Comment:
Shouldn't the same prevValidation reset apply when what you fetched is empty? They are basically the same action so should be handled the same way, right?

Reply:
Absolutely, updated.